### PR TITLE
[lua] Support Luau, a gradually typed language based on Lua 

### PR DESCRIPTION
**Pull request by matthargett**
## Describe the PR

Support Luau, which is an overlay on top of Lua 5.x that shouldn't introduce any ambiguity with legacy codebases.
Tested against ~200KLOC of OSS Lua code (comments in file), and 450KLOC of private Lua code.  I added a new test file and snapshot that captures the scenarios in the aforementioned corpus' that found bugs in my grammar modifications.

CPD now analyzes all of the StyLua and full-moon test suites without errors, except for intentional abuses cases that involve unterminated strings. 🎉 

This also adds in the Java CPD feature to skip imports (require statements in Lua's case), the repetitive boilerplate for which can get in the way of lowering the duplication threshold for logic. 

Additionally, the Lua tokenizer now supports the C# CPD feature to skip literals and literal sequences, so that non-behavioral data tables (arrays of constants for cryptography, translation string tables, etc) can be removed so the duplicate threshold can be further lowered for actual behavioral code.

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">No regression tested rules have been changed.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-lua/src/test/resources/net/sourceforge/pmd/lang/lua/cpd/testdata/luauTypes.txt [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    [,]                                     24        24
+    [...]                                   26        28
+    [:]                                     29        29
+    [boolean]                               31        37
+    [)]                                     38        38
+    [:]                                     39        39
+    [(]                                     41        41
+    [{]                                     42        42
+    [\[]                                    43        43
+    [string|number]                         44        56
```
Comment (oowekyala) [inline pmd-lua/src/test/resources/net/sourceforge/pmd/lang/lua/cpd/testdata/luauTypes.txt:84]: This doesn't look properly tokenized, is it?

File: pmd-lua/src/test/resources/net/sourceforge/pmd/lang/lua/cpd/testdata/luauTypes.txt [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    [\]]                                    57        57
+    [:]                                     58        58
+    [T]                                     60        60
+    [}]                                     61        61
+    [,]                                     62        62
+    [string]                                64        69
+    [,]                                     70        70
+    [Function<...any>]                      72        87
```
Comment (oowekyala) [inline pmd-lua/src/test/resources/net/sourceforge/pmd/lang/lua/cpd/testdata/luauTypes.txt:92]: Same

File: pmd-lua/src/test/resources/net/sourceforge/pmd/lang/lua/cpd/testdata/luauTypes.txt [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    [::]                                    23        24
+    [{]                                     26        26
+    [T]                                     28        28
+    [}]                                     30        30
+    [,]                                     31        31
+    [body]                                  33        36
+    [,]                                     37        37
+    [function]                              39        46
+    [(]                                     47        47
+    [...)]                                  48        51
```
Comment (oowekyala) [inline pmd-lua/src/test/resources/net/sourceforge/pmd/lang/lua/cpd/testdata/luauTypes.txt:231]: Same

File: pmd-lua/src/main/antlr4/net/sourceforge/pmd/lang/lua/antlr4/Lua.g4 [hunk 4]
```diff
@@ -234,6 +245,58 @@ string
     : NORMALSTRING | CHARSTRING | LONGSTRING
     ;
 
+SimpleType
```
Comment (oowekyala) [inline pmd-lua/src/main/antlr4/net/sourceforge/pmd/lang/lua/antlr4/Lua.g4:248]: These are lexer rules... is that on purpose? The issues I pointed out would not show up were this a parser rule I think

**Review by oowekyala — COMMENTED** [review]
Hi and thanks for the contribution!

**Comment by matthargett** [comment]
> Hi and thanks for the contribution! 

You're welcome! I noticed I need to consistently format the grammar with the original coding style. I'll do that today.

**Comment by matthargett** [comment]
Just a heads up that I'm going to be on vacation for 2 weeks. I'll put this PR in draft until I'm back, then fold in the feedback. Thanks!!

**Comment by matthargett** [comment]
Just a few more cleanups, from my perspective, but let me know if there's additional feedback!

File: pmd-lua/src/main/antlr4/net/sourceforge/pmd/lang/lua/antlr4/Lua.g4 (hunk 4 above)
Comment (matthargett) [reply pmd-lua/src/main/antlr4/net/sourceforge/pmd/lang/lua/antlr4/Lua.g4]: derp. fixed.

File: pmd-lua/src/test/resources/net/sourceforge/pmd/lang/lua/cpd/testdata/luauTypes.txt (hunk 3 above)
Comment (matthargett) [reply pmd-lua/src/test/resources/net/sourceforge/pmd/lang/lua/cpd/testdata/luauTypes.txt]: fixed.

File: pmd-lua/src/test/resources/net/sourceforge/pmd/lang/lua/cpd/testdata/luauTypes.txt (hunk 2 above)
Comment (matthargett) [reply pmd-lua/src/test/resources/net/sourceforge/pmd/lang/lua/cpd/testdata/luauTypes.txt]: fixed.

File: pmd-lua/src/test/resources/net/sourceforge/pmd/lang/lua/cpd/testdata/luauTypes.txt (hunk 1 above)
Comment (matthargett) [reply pmd-lua/src/test/resources/net/sourceforge/pmd/lang/lua/cpd/testdata/luauTypes.txt]: fixed.

**Comment by matthargett** [comment]
@adangel I'm not sure the build timeout that causes CI to fail is relate to my changes, but I can't re-run the job. Can you review and debug from your side? Thanks! :D

**Comment by matthargett** [comment]
@adangel I finally won the dice roll with the maven server being overloaded. Would it be possible for this PR to be included in 6.50 release?

**Comment by adangel** [comment]
@matthargett Thanks for the update! I'll try to review it next week - having it in 6.50.0 shouldn't be a problem.

**Review by adangel — APPROVED** [review]
Thanks!

I'm fixing the remaining parts when I merge your PR.

Note: I'm also fixing are parsing issue around comments, mentioned here (https://github.com/antlr/grammars-v4/commit/ff6ef1e876060e9e66535e88879e1ccd045647e4#r84475967). Keep this in mind when updating the grammar back in https://github.com/antlr/grammars-v4/tree/master/lua

One general comment: The grammar now supports both "original" Lua and Luau. This is OK for PMD, where we can be less strict when parsing. But if the upstream grammar is used to verify "original" Lua code and one expects to fail parsing invalid Lua code (but valid Luau), that expectation is now not true anymore. Maybe a separate grammar is better - but it always depends on the use case.

**Comment by matthargett** [comment]
> Note: I'm also fixing are parsing issue around comments, mentioned here ([antlr/grammars-v4@ff6ef1e#r84475967](https://github.com/antlr/grammars-v4/commit/ff6ef1e876060e9e66535e88879e1ccd045647e4#r84475967)). Keep this in mind when updating the grammar back in https://github.com/antlr/grammars-v4/tree/master/lua

Thanks! I saw that person's comment and was meaning to circle back and debug. It's funny, issues with `'` and `"` in comments was one of the issues I fixed in my first PR!

> One general comment: The grammar now supports both "original" Lua and Luau. This is OK for PMD, where we can be less strict when parsing. But if the upstream grammar is used to verify "original" Lua code and one expects to fail parsing invalid Lua code (but valid Luau), that expectation is now not true anymore. Maybe a separate grammar is better - but it always depends on the use case.

Agreed. Here, I tried to see if a superset approach would work across all the OSS-available variants (Lua 5.x, Luau, WoW Lua, and LuaJIT) and strved to ensure they all had representation in the OSS testing corpus (linked in the comment header of the file). It *looks* like everything works when running CPD across those many repos and 400+KLOC (and my additional private 300+KLOC), but it's feasible that someone (a human, not a fuzzing tool) could write new code that would cause issues. If that happens, I'm happy to split the grammar into whatever discrete pieces are helpful for that new use case.

Until then, having a usable duplication checker is an enormous leap forward for teams that are scaling engineer operations around Lua (whatever variant or runtime) -- in either headcount or KLOC dimensions (or both!).

Thanks again for the assist!

**Pull request merged**